packages/astro/src/runtime/server/hydration.ts

```diff
@@ -164,6 +164,15 @@ export async function generateHydrateScript(
 
 	island.props['ssr'] = '';
 	island.props['client'] = hydrate;
+
+	if (renderer.name === '@astrojs/react' && metadata.reactServerActionResult) {
+		const { key, name, value } = metadata.reactServerActionResult;
+
+		island.props['action-result'] = JSON.stringify(value);
+		island.props['action-key'] = key;
+		island.props['action-name'] = name;
+	}
```

Comment:
A renderer can add extra attributes that will be attached to the client. This is how Preact can have shared signals. See the server code here: https://github.com/withastro/astro/blob/0dbd8eeb77065f3ed03f481c8042f2896a5448c4/packages/integrations/preact/src/server.ts#L94

And here's how they are retrieved on the client: https://github.com/withastro/astro/blob/0dbd8eeb77065f3ed03f481c8042f2896a5448c4/packages/integrations/preact/src/client.ts#L18-L29

Reply:
Huh, that's interesting! I worry that's not powerful enough for what we're doing here though. We need to pass new information from the `SSRResult` that we previously did not track. Namely, the request `formData`. I _could_ make this generic, but it felt like a premature abstraction when we don't know how or if other integrations could take advantage of it. Hence the very strict name `reactServerActionResult`. If y'all would prefer I pass down the request formData without any preprocessing, and expose a new helper from the integration, I could do so.

Reply:
Why does that information need to be passed down? Why can't the React renderer get it when it needs it?

Reply:
@matthewp Because we do not expose the `request` from the `SSRResult` at all. So there's no way for renderers to access the request body. React needs the request body to parse the `formData` to determine what `form` should update for progressive enhancemenet.

Reply:
Ok, so could we pass the request through the SSRResult instead? Would that satisfy the need? The renderer could get the request, use the SSRResult as a key so that it knows it has only done it once.

Reply:
@matthewp Yes, I think passing the `request` would be perfect 👍 Implemented